pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,88 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter-core/pkg/utils/resources"
+)
+
+const pollingPeriod = 5 * time.Second
+
+type Controller struct {
+	cluster *state.Cluster
+}
+
+func NewController(cluster *state.Cluster) *Controller {
+	return &Controller{
+		cluster: cluster,
+	}
+}
+
+func (c *Controller) Name() string {
+	return "metric_scraper"
+}
+
+func (c *Controller) Builder(_ context.Context, mgr manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(mgr)
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// Reset all the gauges ahead of setting the new gauges with the new labels
+	for _, v := range gauges {
+		v.Reset()
+	}
+	// Populate metrics
+	c.cluster.ForEachNode(func(n *state.StateNode) bool {
```

Comment:
This holds the lock as long as it is completing it's function. The only difference between this and `cluster.Nodes()` would be:
1. `cluster.Nodes()` deepcopies all the nodes while the `ForEachNode` doesn't
2. `cluster.Nodes()` _might_ hold the lock for a slightly smaller amount of time because it's possible that the `DeepCopy` is more efficient than the iteration function, but that might not also be true.

In general, I've been using `cluster.Nodes()` when we are _already_ needing to do a deep copy so it only increases our efficiency. Here, I don't think we need the deep-copy so it's less clear if we should do one or the other

Reply:
Given how hot cluster state is, we should minimize locks on it whereever possible.

Reply:
I'm not sure if I get your point here. Which thing are you advocating for? Both of them will hold a lock and they have to hold a lock (at least briefly) to get details on the cluster state.